### ISSUE TITLE
Performance Reproduction Problem

### ISSUE THREAD

**Issue by eeeric-code** [niki-amini-naieni/CountGD#16]
Hi, it is a great work!
When reproducing this project, I am able to achieve performance close to that reported in the paper on the FSC147 val set. However, when testing on the FSC147 test set using the same checkpoint, we obtain MAE≈11 and RMSE≈100 (sam tt-norm=False, crop=False, remove-bad-exemplar=False), which is close to the paper's results MAE=10.92 and RMSE=99.58. However, we only observe MAE=7+ and RMSE=80+ (sam tt-norm=True, crop=True, remove-bad-exemplar=True), which is different from the paper's results MAE=5.74 and RMSE=24.09. Could you please advise on the potential reasons for this discrepancy? Despite some of the reproduction results are consistent.

**Comment by niki-amini-naieni** [comment]
Hi thank you for your question. Is this for the pretrained checkpoint, or are you retraining the model?

**Comment by niki-amini-naieni** [comment]
The results for the pretrained checkpoint should be identical to the paper results

**Comment by eeeric-code** [comment]
I have retrained the model

**Comment by eeeric-code** [comment]
With the retrained checkpoint, the results are close to paper's results when sam tt-norm=True, crop=True, remove-bad-exemplar=True. But results are different from paper when sam tt-norm=False, crop=False, remove-bad-exemplar=False. That's weird.

**Comment by niki-amini-naieni** [comment]
Ah okay, got it. I have not released the training code yet, so I am not able to reproduce your results on my side. I, however, can still speculate. This may be due to a high variance caused by a couple examples in the test set with very high counts driving up the rmse. You can check this by omitting examples with greater than 900 objects when you calculate the test error. If this results in a significant improvement to the error, then this is probably the issue. To improve the robustness of the training code to this issue, you could apply the adaptive cropping to the early stopping code so that when the validation set error is evaluated during early stopping, adaptive cropping is applied. Right now, there is some other source of non-determinism, other than the seed, in the posted code, and adaptive cropping is not being applied during early stopping.

**Comment by eeeric-code** [comment]
Thanks a lot! I will try it.

**Comment by niki-amini-naieni** [comment]
RE: With the retrained checkpoint, the results are close to paper's results when sam tt-norm=True, crop=True, remove-bad-exemplar=True. But results are different from paper when sam tt-norm=False, crop=False, remove-bad-exemplar=False. That's weird.

The main results in the papper have sam_tt_norm=True and remove_bad_exemplar=True, so it is not weird. We report results without these options in the appendix pasted below:
![image](https://github.com/user-attachments/assets/36b1794a-116f-41c7-9273-4b8a3cbd3d97)

**Comment by niki-amini-naieni** [comment]
The influence of these options is described in the appendix here:
![image](https://github.com/user-attachments/assets/9e383030-0e2c-49f7-aec5-5fccb7af2505)
![image](https://github.com/user-attachments/assets/899271cd-36a1-423a-8d72-bf29a29846ea)

**Comment by eeeric-code** [comment]
> RE: With the retrained checkpoint, the results are close to paper's results when sam tt-norm=True, crop=True, remove-bad-exemplar=True. But results are different from paper when sam tt-norm=False, crop=False, remove-bad-exemplar=False. That's weird.
> 
> The main results in the papper have sam_tt_norm=True and remove_bad_exemplar=True, so it is not weird. We report results without these options in the appendix pasted below: 

Sorry, I make a mistake in my previous response. It should be:  **With the retrained checkpoint, the results are close to paper's results when sam tt-norm=False, crop=False, remove-bad-exemplar=False. But results are different from paper when sam tt-norm=True, crop=True, remove-bad-exemplar=True.**

**Comment by eeeric-code** [comment]
> The influence of these options is described in the appendix here: 

I have noticed that and conducted some ablation study, but still can not find the factors affecting the results

**Comment by eeeric-code** [comment]
Thanks! Let me check.

**Comment by ShiZiqiang** [comment]
> With the retrained checkpoint, the results are close to paper's results when sam tt-norm=True, crop=True, remove-bad-exemplar=True. But results are different from paper when sam tt-norm=False, crop=False, remove-bad-exemplar=False. That's weird.

Hi, Eric.  I am also retraining CountGD checkpoint, using the following command

python -u main_train_study.py --output_dir ./countgd_train -c config/cfg_fsc147_test.py --datasets config/datasets_fsc147_test.json --options text_encoder_type=/data5/shi/visual_counting/CountGD/CountGD-main/checkpoints/bert-base-uncased

I would like to ask what command you used? Or can you share your training script? Thank you.

**Comment by eeeric-code** [comment]
The same as yours, except for additionally using a pretrained GroundingDINO from https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha2/groundingdino_swinb_cogcoor.pth
Just add another parameter: --pretrain_model_path=./checkpoints/groundingdino_swinb_cogcoor.pth

**Comment by ShiZiqiang** [comment]
> The same as yours, except for additionally using a pretrained GroundingDINO from https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha2/groundingdino_swinb_cogcoor.pth Just add another parameter: --pretrain_model_path=./checkpoints/groundingdino_swinb_cogcoor.pth

Thank you so much. It seems that my understanding is correct. Thank you for your suggestion. No wonder my results are so bad, it turns out that I did not use the pre-trained GroundingDINO. Thank you.

**Comment by David-19940718** [comment]
Hi, @ShiZiqiang;

I’m delighted to hear that you’ve managed to successfully replicate the project. I was wondering, did you happen to use a single machine with just one GPU for the reproduction? I attempted to leverage multiple GPUs on a single machine, but unfortunately, I encountered some issues.

**Comment by ShiZiqiang** [comment]
> Hi, @ShiZiqiang;
> 
> I’m delighted to hear that you’ve managed to successfully replicate the project. I was wondering, did you happen to use a single machine with just one GPU for the reproduction? I attempted to leverage multiple GPUs on a single machine, but unfortunately, I encountered some issues.

I used a single GPU on a single machine. I haven't reproduced it completely yet, still struggling. Eric @eeeric-code seems to have reproduced some of the effects. We can exchange reproduction experience here :).

**Comment by eeeric-code** [comment]
A single GPU is enough for me. Multi-GPUs can be easily applied with DDP as the project already contains the necessary code.
